cmd/maya-apiserver/spc-actions/select_disk.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2017 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storagepoolactions
+
+import (
+	mach_apis_meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	//openebs "github.com/openebs/maya/pkg/client/clientset/versioned"
+	"errors"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	openebs "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+	"github.com/openebs/maya/pkg/client/k8s"
+)
+
+// clientset struct holds the interface of internalclientset
+// i.e. openebs.
+// This struct will be binded to method ListDisk and is useful in mocking
+// and unit testing.
+type clientSet struct {
+	oecs openebs.Interface
+}
+
+// nodeDisk struct will be used as a value for a map nodeDiskMap (map defined in ListDisk function)
+// The struct will be useful in forming the data structure nodeDiskMap which will be manipulated
+// to efficiently select the nodes and disk for dynamic pool provisioning.
+type nodeDisk struct {
+	// diskCount is the count of usable disks that can be used in storagepool provisioning.
+	diskCount int
+	//diskList is the list of usable disks that can be used in storagepool provisioning.
+	diskList []string
+}
+
+// spareAllotment holds the value for the remaining node allotments
+// for pool provisioning.
+var spareAllotment int16
+
+// getDiskList is a wrapper function which will receive list of disks
+// that can be used for dynamic pool provisioning at runtime.
+// The function finally returns the disk list to the caller (i.e. getCasPoolDisk function).
+func getDiskList(cp *v1alpha1.CasPool) ([]string, error) {
+
+	// Get kubernetes clientset
+	// namespaces is not required, hence passed empty.
+	newK8sClient, err := k8s.NewK8sClient("")
+
+	if err != nil {
+		return nil, err
+	}
+	// Get openebs clientset using a getter method (i.e. GetOECS() ) as
+	// the openebs clientset is not exported.
+	newOecsClient := newK8sClient.GetOECS()
+
+	// Create instance of clientset struct defined above which binds
+	// ListDisk method and fill it with openebs clienset (i.e.newOecsClient ).
+	newClientSet := clientSet{
+		oecs: newOecsClient,
+	}
+
+	// if no minimum pools were specified it will default to 1.
+	if cp.MinPools <= 0 {
+		glog.Warning("invalid or 0 min pool specified, defaulting to 1")
+		cp.MinPools = 1
```

Comment:
Thinking more on this. Will use of different strategies eliminate the use of min & max concept. How about implementing a fallback process where a list of strategies can be specified in a specified order and logic tries to implement the suitable strategy in order.

e.g.
```yaml
poolSelector:
  - name: openebsSelector
  - name: valueSelector
    value: 2
  - name: valueSelector
    value: 1
```

Reply:
The value needs to be a string that can be un-marshalled into different structures based on the name of the pool selector.

Reply:
Any reason why code level defaults should not be set ? Here one can any way override.
I do not want to fail the creation of of storagepool if `minPool` is not specified.
By doing it other way via env etc... what if user does not provide that field(`minPool`)?
If I am trying to keep a field optional , and it is used then it will get defaulted to some value.. Right??
Are there any other way of defaulting it ? This is what came to my head while coding it.
@AmitKumarDas

Reply:
The strategy thing , I am not able to see why it should be here. @AmitKumarDas